src/utils/jump-addon.js

```diff
@@ -0,0 +1,150 @@
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import CodeMirror from 'codemirror';
+
+CodeMirror.defineOption('jump', false, function (cm, options, old) {
+  if (old && old !== CodeMirror.Init) {
+    const oldOnMouseOver = cm.state.jump.onMouseOver;
+    CodeMirror.off(cm.getWrapperElement(), 'mouseover', oldOnMouseOver);
+    const oldOnMouseOut = cm.state.jump.onMouseOut;
+    CodeMirror.off(cm.getWrapperElement(), 'mouseout', oldOnMouseOut);
+    CodeMirror.off(document, 'keydown', cm.state.jump.onKeyDown);
+    delete cm.state.jump;
+  }
+
+  if (options) {
+    const state = cm.state.jump = {
+      options,
+      onMouseOver: onMouseOver.bind(null, cm),
+      onMouseOut: onMouseOut.bind(null, cm),
+      onKeyDown: onKeyDown.bind(null, cm),
+    };
+
+    CodeMirror.on(cm.getWrapperElement(), 'mouseover', state.onMouseOver);
+    CodeMirror.on(cm.getWrapperElement(), 'mouseout', state.onMouseOut);
+    CodeMirror.on(document, 'keydown', state.onKeyDown);
+  }
+});
+
+function onMouseOver(cm, event) {
+  const target = event.target || event.srcElement;
+  if (target.nodeName !== 'SPAN') {
+    return;
+  }
+
+  const box = target.getBoundingClientRect();
+  const cursor = {
+    left: (box.left + box.right) / 2,
+    top: (box.top + box.bottom) / 2
+  };
+
+  cm.state.jump.cursor = cursor;
+
+  if (cm.state.jump.isHoldingModifier) {
+    enableJumpMode(cm);
+  }
+}
+
+function onMouseOut(cm) {
+  if (!cm.state.jump.isHoldingModifier && cm.state.jump.cursor) {
+    cm.state.jump.cursor = null;
+    return;
+  }
+
+  if (cm.state.jump.isHoldingModifier && cm.state.jump.marker) {
+    disableJumpMode(cm);
+  }
+}
+
+function onKeyDown(cm, event) {
+  if (cm.state.jump.isHoldingModifier || !isJumpModifier(event.key)) {
+    return;
+  }
+
+  cm.state.jump.isHoldingModifier = true;
+
+  if (cm.state.jump.cursor) {
+    enableJumpMode(cm);
+  }
+
+  const onKeyUp = upEvent => {
+    if (upEvent.code !== event.code) {
+      return;
+    }
+
+    cm.state.jump.isHoldingModifier = false;
+
+    if (cm.state.jump.marker) {
+      disableJumpMode(cm);
+    }
+
+    CodeMirror.off(document, 'keyup', onKeyUp);
+    CodeMirror.off(document, 'click', onClick);
+    cm.off('mousedown', onMouseDown);
+  };
+
+  const onClick = clickEvent => {
+    const destination = cm.state.jump.destination;
+    if (destination) {
+      cm.state.jump.options.onClick(destination, clickEvent);
+    }
+  };
+
+  const onMouseDown = (_, downEvent) => {
+    if (cm.state.jump.destination) {
+      downEvent.codemirrorIgnore = true;
+    }
+  };
+
+  CodeMirror.on(document, 'keyup', onKeyUp);
+  CodeMirror.on(document, 'click', onClick);
+  cm.on('mousedown', onMouseDown);
+}
+
+const isMac = navigator && navigator.appVersion.indexOf('Mac') !== -1;
+const isWin = navigator && navigator.appVersion.indexOf('Win') !== -1;
+
+function isJumpModifier(key) {
+  return !isWin && key === 'Meta' || !isMac && key === 'Control';
```

Comment:
Not sure what `navigator.appVersion` returns on Linux, but I would expect it's probably customary to use Control for this purpose there, in which case, this would want to be something like:

```
return isMac && key === 'Meta' || !isMac && key === 'Control';
```